---
title: "cl13"
output: html_document
---

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(dplyr)
library(devtools)
library(Matrix)
library(data.table)
library(cowplot)
library(RColorBrewer)
library(scales)
library(reshape2)
library(dplyr)
#library(future)
library(tidyverse)
library(SeuratData)
library(harmony)
library(ggrepel)
library(nichenetr)
#plan(multisession, workers=24)
```

#load and preprocess datasets from cellranger multi 
```{r}
directobj <- function(hkc,sample){
  object <- Read10X(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/01_preprocess/runs/",hkc,"/outs/per_sample_outs/",sample,"/count/sample_filtered_feature_bc_matrix"))
  obj <- CreateSeuratObject(counts = object$`Gene Expression`, assay = "RNA",project = str_c(hkc,sample))
  obj[["ADT"]] <- CreateAssayObject(counts = object$`Antibody Capture`)
  obj[["GDO"]] <- CreateAssayObject(counts = object$`CRISPR Guide Capture`)
  obj[["HTO"]] <- CreateAssayObject(counts = object$`Multiplexing Capture`)
  obj
}

#create obj
hkc <- c("HKC074","HKC075", "HKC076")
sample <- c("CL13_rep1","CL13_rep2","CL13_rep3")
cl13 <- map2(hkc,sample,directobj)
names(cl13) <- str_c(hkc,"_",sample)

#merge
cl13 <- merge(x = cl13[[1]],
		       y = cl13[2:length(cl13)],
		       merge.data = TRUE,add.cell.ids = names(cl13)) 

#quality control
cl13[["percent.mt"]] <- PercentageFeatureSet(cl13,pattern="^mt") 
cl13[["percent.ribo"]] <- PercentageFeatureSet(cl13,pattern = "^Rp[sl][[:digit:]]|^RP[SL][[:digit:]]")
VlnPlot(cl13, features = c("percent.mt","percent.ribo","nFeature_RNA","nCount_RNA"),split.by = "orig.ident",ncol=2)
cl13_filtered <- subset(cl13,subset = (percent.mt<5) & (nFeature_RNA>200) & (percent.ribo>5))
```

# threshold  from cellranger aggr
get threshold from cellranger multi 
/data/msun/projects/kay_ECCITE/23TP_merged/01_preprocess/runs/cl13_nonorm/outs/count/crispr_analysis/protospacer_umi_thresholds.csv
```{r}
gdo <- as.data.frame(cl13_filtered@assays$GDO@counts) 
threshold <- read.csv("/data/msun/projects/kay_ECCITE/23TP_merged/01_preprocess/runs/cl13_nonorm/outs/count/crispr_analysis/protospacer_umi_thresholds.csv") 
threshold$Protospacer <- gsub("_","-",threshold$Protospacer)
threshold$Protospacer[1:3] <- c("gScramble1-1","gScramble3-2","gScramble5-4")
gdo <- gdo[rownames(gdo) %in% threshold$Protospacer,]
#gdo value of cells under threshold is 0, otherwise is 1
for (i in 1:nrow(gdo)) {
  n <- threshold[threshold$Protospacer == rownames(gdo)[i],]$UMI.threshold # Extract the threshold value using subsetting
  gdo[i, ] <- ifelse(gdo[i,] < n, 0, 1)#grna<n ~ 0, grna>=n ~ 1
}
#subset single perturbed cells: colsum(gdo)=1
gdo_single <- gdo[, which(colSums(gdo == 1) == 1)] # cells with single grna perturbed
singlegrna<- data.frame(apply(gdo_single, 2, function(x) rownames(gdo_single)[which(x == 1)])) #call the sgRNA name with value =1
colnames(singlegrna) <- "perturbed_gene"

#multiple perturbed cells
gdo_multi <- gdo[, which(colSums(gdo == 1) > 1)] #  cells with more than 1 gRNA

#add perturbed_gene and guide_ID and perturbed_ID column
#perturbed_gene: gScramble, gZscan20
#perturbed_ID: gScramble, gZscan20-4
#guide_ID:gScramble-1,gZscan20-4
metadata <- cl13_filtered@meta.data %>% merge(.,singlegrna, by = "row.names",all.x=TRUE) %>% column_to_rownames(var="Row.names")
metadata$guide_ID <- metadata$perturbed_gene
metadata$perturbed_ID <- case_when(metadata$guide_ID %in% c("gScramble3-2","gScramble5-4","gScramble1-1") ~ "gScramble",
                                       TRUE ~ metadata$guide_ID)
metadata$perturbed_gene <- case_when(metadata$guide_ID %in% c("gScramble3-2","gScramble5-4","gScramble1-1") ~ "gScramble",
                                     rownames(metadata) %in% colnames(gdo_multi) ~ "multi_gRNA",
                                       TRUE ~ substr(metadata$guide_ID,1,nchar(metadata$guide_ID)-2))
metadata[is.na(metadata$perturbed_gene),]$perturbed_gene <- "no_gRNA"

#add perturbed_status column
metadata <- metadata %>% mutate(perturbed_status = case_when(perturbed_gene == "multi_gRNA" ~ "Multi_perturbed",
                                                             perturbed_gene == "no_gRNA" ~ "No_perturbed",
                                                             TRUE ~ 'Single_perturbed'))

cl13_filtered@meta.data <- metadata
cl13_single <- subset(cl13_filtered, subset = perturbed_status == "Single_perturbed")
saveRDS(cl13_single,file = "/data/msun/projects/kay_ECCITE/23TP_merged/cl13_single.rds")
```


#split bam file based on perturbed_gene
```{r}
cl13_cb_gRNA <- cl13_single@meta.data[cl13_single$guide_ID %in% c("gNfil3-2","gNfil3-4","gZfp324-2","gZfp324-4","gStat3-2","gStat3-4","gHic1-2","gHic1-4","gIkzf3-2","gIkzf3-4","gPrdm1-2","gPrdm1-4","gZscan20-2","gZscan20-4","gJdp2-2","gJdp2-4","gScramble"),c("guide_ID","perturbed_gene")] %>%
  mutate(barcodes=substr(rownames(.),18,nchar(rownames(.)))) %>%
  .[,c(3,1)]
write.table(cl13_cb_gRNA,file = "/data/msun/projects/kay_ECCITE/23TP_merged/02_koeff/gRNA_cb.txt",quote=FALSE,row.names = FALSE,col.names = FALSE,sep = "\t")
```

#preprocess
```{r}
DefaultAssay(cl13_single) <- "RNA"
cl13_single_normscale <- cl13_single %>%
  NormalizeData() %>%
  FindVariableFeatures(selection.method = "vst",nfeatures = 2000,verbose = FALSE) %>%
  ScaleData() %>%
  RunPCA(npcs = 30, verbose = FALSE) %>%
  RunUMAP(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
  FindNeighbors(reduction = "pca", dims = 1:30, verbose = FALSE) %>%
  FindClusters(verbose = FALSE,resolution=c(seq(.1,1,.1)))
saveRDS(cl13_single_normscale,file = "/data/msun/projects/kay_ECCITE/23TP_merged/cl13_single_normscale.rds")
```

#labeled obejct from Kay
```{r}
cl13_label <- readRDS("/data/hkchung/23-TP0304/merged_cleaned_TP04_230802.rds")

plot <- DimPlot(cl13_label, reduction = "umap")
select.cells <- CellSelector(plot = plot)
Idents(cl13_label, cells = select.cells) <- "SubCells"
sub_cells <- WhichCells(cl13_label, idents = "SubCells")
sub_cl13 <- subset(cl13_label, cells = sub_cells)
saveRDS(sub_cl13,file = "/data/msun/projects/kay_ECCITE/23TP_merged/sub_cl13.rds")

Idents(sub_cl13) <- "Cluster"
DefaultAssay(sub_cl13) <- "RNA"
DimPlot(sub_cl13)
```

#summary table 
guide_ID: gZscan20-2 gScramble1-2
perturbed_gene: Zscan20 gScramble
```{r}
cellnumebr <- sub_cl13@meta.data %>% 
  group_by(guide_ID,Cluster,orig.ident) %>% 
  summarize(cellnumber = n()) %>%
  spread(.,Cluster,cellnumber)
write.csv(cellnumebr, file = "/data/msun/projects/kay_ECCITE/23TP_merged/02_koeff/guideID_cellnum_cl13.csv")
```

#density plot of gRNA
```{r}
library(gridExtra)

#loop in every sgrna
library("Nebulosa")
for (grna in rownames(sub_cl13@assays$GDO)) {
  tryCatch({
    DefaultAssay(sub_cl13) <- "GDO"
    p <- plot_density(sub_cl13,grna)
    ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/03_analysis/density plots/",grna,".pdf"),width = 6, height = 5)
  }, error=function(e){})
}

#plots of selected sgrna
plot_density(sub_cl13, c("gScramble3-2", "gScramble5-4","gZscan20-2","gZscan20-4","gHic1-2","gHic1-4","gStat3-2","gStat3-4","gJdp2-2","gJdp2-4","gPrdm1-2","gPrdm1-4","gIkzf3-2","gIkzf3-4","gZfp324-2","gZfp324-4","gNfil3-2","gNfil3-4")) + plot_layout(ncol = 6)
```

#density plot of gRNA in sub_cl13 without cellcyle
change pertubed_ID and list1 each time 
```{r}
library(cetcolor)
list1 <- unique(cl13_nocellcyle$perturbed_ID)
list2 <- unique(cl13_nocellcyle$perturbed_gene) 
for (grna in list1) {
  tryCatch({
    scale.col <- cet_pal(16, name = "fire")
    sgrna_cells <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_ID == grna,])
    data <- Embeddings(object = cl13_nocellcyle[["umap"]])[sgrna_cells,c(1,2)] %>% as.data.frame()
    p <- DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.3, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = scale.col) +
      ggtitle(grna)

    ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/03_analysis/densityplot without cellcycle_identity/",grna,".pdf"),width = 6, height = 5)
  }, error=function(e){})
}

for (grna in list2) {
  tryCatch({
    scale.col <- cet_pal(16, name = "fire")
    sgrna_cells <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_gene == grna,])
    data <- Embeddings(object = cl13_nocellcyle[["umap"]])[sgrna_cells,c(1,2)] %>% as.data.frame()
    p <- DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.3, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = scale.col) +
      ggtitle(grna)

    ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/03_analysis/densityplot without cellcycle_identity/",grna,".pdf"),width = 6, height = 5)
  }, error=function(e){})
}
```

#density plot of three groups of perturbed_gene
```{r}
group1 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_gene %in% c("gPrdm1","gNfatc1","gStat3"),])
group2 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_gene %in% c("gIkzf3","gJdp2","gNr4a2","gZscan20","gHic1"   ,"gZfp410","gArid3a","gEtv5","gNfil3","gHinfp","gFoxd2","gStat3"  ,"gGfi1","gIrf8","gZbtb49","gPrdm4","gZfp324","gTfdp1","gZfp143"),])
group3 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_gene %in% c("gScramble"),])

data1 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group1,c(1,2)] %>% as.data.frame()
data2 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group2,c(1,2)] %>% as.data.frame()
data3 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group3,c(1,2)] %>% as.data.frame()

color1 <- cet_pal(16, name = "fire")
color2 <- cet_pal(16, name = "l16")
color3 <- cet_pal(16, name = "d6")

DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data1,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = color1) 

DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data2,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(2, 2)) +
      scale_fill_gradientn(colours = color2) 

DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data3,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = color1) 
```

#density plot of three groups of perturbed_ID
```{r}
group1 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_ID %in% c("gNfatc1-4","gStat3-4","gPrdm1-2","gIkzf3-2","gHic1-4","gZfp324-4","gIrf8-4","gNr4a2-2","gZfp143-4","gGfi1-2","gNfil3-2","gPrdm4-2","gJdp2-2","gZscan20-4","gFoxd2-2","gZfp410-2","gTfdp1-2","gZbtb49-4","gArid3a-2","gEtv5-4","gHinfp-4"),])
group1 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_ID %in% c("gNfatc1-4","gStat3-4","gPrdm1-2"),])
group2 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_ID %in% c("gIkzf3-2","gHic1-4","gZfp324-4","gIrf8-4","gNr4a2-2","gZfp143-4","gGfi1-2","gNfil3-2","gPrdm4-2","gJdp2-2","gZscan20-4","gFoxd2-2","gZfp410-2","gTfdp1-2","gZbtb49-4","gArid3a-2","gEtv5-4","gHinfp-4"),])
group3 <- rownames(cl13_nocellcyle@meta.data[cl13_nocellcyle$perturbed_gene %in% c("gScramble"),])

data1 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group1,c(1,2)] %>% as.data.frame()
data2 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group2,c(1,2)] %>% as.data.frame()
data3 <- Embeddings(object = cl13_nocellcyle[["umap"]])[group3,c(1,2)] %>% as.data.frame()

color1 <- cet_pal(16, name = "fire")
color2 <- cet_pal(16, name = "l16")
color3 <- cet_pal(16, name = "d6")

p1 <- DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data1,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = color1) 

p2 <- DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data2,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(4, 4)) +
      scale_fill_gradientn(colours = color1) 

p3 <- DimPlot(cl13_nocellcyle,cols = c("#F4C5CC", "#73AACF", "#9FB27F")) +
      stat_density_2d(data3,mapping=aes(x = UMAP_1, y = UMAP_2, fill = after_stat(level)), 
                  linewidth = 0.1, geo = "density_2d_filled", 
                  colour = "ivory", alpha = 0.2, n = 150, h = c(1.2, 1.2)) +
      scale_fill_gradientn(colours = color1) 


```

#PHATE
# pip install phate
```{r}
reticulate::py_discover_config(required_module = "phate")
reticulate::import("phate")
library(phateR)
  library(BiocFileCache)
  library(scran)
  library(scater)
  library(tidyverse)
  library(SingleCellExperiment)
  library(DropletUtils)
  library(ggplot2)
  library(cowplot)
  library(Seurat)
  library(monocle3)
  library(SeuratObject)
  library(viridis)
set.seed(11)
seurat_data <- as.data.frame(sub_cl13@assays$RNA@data) %>% t(.)
phate_output <- phate(seurat_data)  #run phate
ggplot(phate_output, aes(x=PHATE1, y=PHATE2)) + geom_point() ## quick sanity check of the phate embeddings
sub_cl13[["PHATE"]] <- CreateDimReducObject(embeddings = phate_output$embedding, key = "PHATE_", assay = "RNA")
DimPlot(sub_cl13, reduction = "PHATE") + ggtitle(label = "PHATE")
```

#PHATE analysis on selected gRNA
run on R4.2
remove cell cycle clusters
```{r}
cl13_nocellcyle <- subset(sub_cl13,subset=Cluster=="Cellcycle", invert=TRUE)
seurat_data <- as.data.frame(cl13_nocellcyle@assays$RNA@data) %>% t(.)
phate_output <- phate(seurat_data)  #run phate
cl13_nocellcyle[["PHATE"]] <- CreateDimReducObject(embeddings = phate_output$embedding, key = "PHATE_", assay = "RNA")
saveRDS(cl13_nocellcyle ,file = "/data/msun/projects/kay_ECCITE/23TP_merged/06_monocle3/cl13_nocellcyle.rds")
cl13_nocellcyle <- readRDS("/data/msun/projects/kay_ECCITE/23TP_merged/06_monocle3/cl13_nocellcyle.rds")
```

#monocle3
need to run on R4.2 
Error in ..Internal(is.unsorted(x, FALSE, FALSE)) : 
  3 arguments passed to .Internal(is.unsorted) which requires 2
```{r}
library(monocle3)
library(SeuratWrappers)
my.so<-cl13_nocellcyle
my.so <- ProjectDim(my.so, reduction = "pca")
expression_matrix <- my.so@assays$RNA@counts

# Get cell metadata
cell_metadata <- my.so@meta.data
if (all.equal(colnames(expression_matrix), rownames(cell_metadata))) {
  print(sprintf("Cell identifiers match"))
} else {
  print(sprintf("Cell identifier mismatch - %i cells in expression matrix, %i cells in metadata",
                ncol(expression_matrix), nrow(cell_metadata)))
  print("If the counts are equal, sort differences will throw this error")
}
gene_annotation <- data.frame(gene_short_name = rownames(my.so@assays$RNA), row.names = rownames(my.so@assays$RNA))
if (all.equal(rownames(expression_matrix), rownames(gene_annotation))) {
  print(sprintf("Gene identifiers all match"))
} else {
  print(sprintf("Gene identifier mismatch - %i genes in expression matrix, %i gene in gene annotation",
                nrow(expression_matrix), nrow(gene_annotation)))
  print("If the counts are equal, sort differences will throw this error")
}

cds <- new_cell_data_set(expression_matrix,
                            cell_metadata = cell_metadata,
                            gene_metadata = gene_annotation)
saveRDS(cds,file = "/data/msun/projects/kay_ECCITE/23TP_merged/06_monocle3/cds.cl13_nocellcycle.rds")
cds <- readRDS("/data/msun/projects/kay_ECCITE/23TP_merged/06_monocle3/cds.cl13_nocellcycle.rds")

#analysis
fData(cds)
fData(cds)$gene_short_name <- rownames(fData(cds))
recreate.partitions <- c(rep(1, length(cds@colData@rownames)))
names(recreate.partitions) <- cds@colData@rownames
cds@clusters@listData[["UMAP"]][["partitions"]] <- as.factor(recreate.partitions)
cds@clusters@listData[["UMAP"]][["clusters"]] <- cl13_nocellcyle$Cluster
cds@int_colData@listData[["reducedDims"]]@listData[["UMAP"]] <- cl13_nocellcyle@reductions$umap@cell.embeddings
cds <- learn_graph(cds, use_partition = F)
p1 <- plot_cells(cds, color_cells_by = "cluster", label_groups_by_cluster = F,
           label_branch_points = T, label_roots = T, label_leaves = F,
           group_label_size = 5)
cds <- order_cells(cds, reduction_method = "UMAP", root_cells = colnames(cds[, clusters(cds) == "Progenitor"]))
p2 <- plot_cells(cds, color_cells_by = "pseudotime", label_groups_by_cluster = T,
           label_branch_points = T, label_roots = F, label_leaves = F)
p1+p2

cl13_nocellcyle$pseudotime <- pseudotime(cds)
saveRDS(cl13_nocellcyle,file = "/data/msun/projects/kay_ECCITE/23TP_merged/06_monocle3/cl13_nocellcyle.rds")

FeaturePlot(cl13_nocellcyle, features = "pseudotime")


```

#dotplot
```{r}
library(scCustomize)
genes <- c("Pdcd1","Cd160","Cd101","Cd244","Tigit","Cd38","Entpd1","Id2","Tox","Bhlhe41","Prdm1","Fil1","Ccr7","Il7r","Sell","Cd27","Cd28","Tcf7","Gzma","Gzmb","Prf1")
Idents(sub_cl13) <- "perturbed_gene"
DotPlot_scCustom(seurat_object = sub_cl13, features = genes, flip_axes = T,
    x_lab_rotate = TRUE,colors_use = viridis_light_high)

Idents(sub_cl13) <- "guide_ID"
DotPlot_scCustom(seurat_object = sub_cl13, features = genes, flip_axes = T,
    x_lab_rotate = TRUE,colors_use = viridis_light_high)

Idents(sub_cl13) <- "orig.ident"
DotPlot_scCustom(seurat_object = sub_cl13, features = sub_cl13, flip_axes = T,
    x_lab_rotate = TRUE,colors_use = viridis_light_high)
```

#scMageck on interested genes
```{r}
library(scMAGeCK)
# 1. RDS can be a Seurat object or local RDS file path that contains the scRNA-seq dataset
DefaultAssay(sub_cl13) <- "RNA"
cl13_allgene <- ScaleData(sub_cl13, features = rownames(sub_cl13))

#2. barcode info
info <- cl13_allgene@meta.data
barcode_rec <- data.frame(cell=rownames(info),barcode=info$guide_ID,sgrna=info$guide_ID,gene=substr(info$perturbed_gene,2,nchar(info$perturbed_gene)),read_count=info$nCount_GDO,umi_count=info$nFeature_GDO)
write.table(barcode_rec,file = "/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/tp04_cl13_normscale_barcode.txt",sep = "\t",quote = FALSE,row.names = FALSE)
barcode="/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/tp04_cl13_normscale_barcode.txt"

#run rra
rraplot <- function(grna){
  rra_cl13 <- scmageck_rra(BARCODE=barcode, RDS=cl13_allgene, GENE=grna,   
                            NEGCTRL="Scramble", KEEPTMP=FALSE, PATHWAY=FALSE) %>%
              mutate(top10_low = rank(-log10(lo_value.low)) %in% 12:21) %>%
              mutate(top10_high = rank(-log10(lo_value.high)) %in% 12:21)
  p1 <- ggplot(data=rra_cl13,aes(x=reorder(Row.names,lo_value.low), y=-log10(lo_value.low),group=1)) +
    geom_line() +
    ggthemes::theme_few() +
    geom_point(data = rra_cl13[rra_cl13$FDR.low<0.05,],color="gray",size=2) +
    geom_point(data = rra_cl13[rra_cl13$top10_low,], color="#4197D9",size=2) +
    ggrepel::geom_text_repel(aes(label = ifelse(top10_low, as.character(Row.names), "")), size = 4,color="#4197D9") +
    xlab("cells with targeted gene") +
    ylab("-log10(RRA score)") +
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
    ggtitle(str_c("Gene KO enriched in ",grna, " low"))
  p2 <- ggplot(data=rra_cl13,aes(x=reorder(Row.names,lo_value.high), y=-log10(lo_value.high),group=1)) +
    geom_line() +
    ggthemes::theme_few() +
    geom_point(data = rra_cl13[rra_cl13$top10_high,], color="orange",size=2) +
    ggrepel::geom_text_repel(aes(label = ifelse(top10_high, as.character(Row.names), "")), size = 4,color="orange") +
    xlab("cells with targeted gene") +
    ylab("-log10(RRA score)") +
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
     ggtitle(str_c("Gene KO enriched in ",grna, " high"))
  p3 <- gridExtra::grid.arrange(p1, p2, ncol = 2)
  ggsave(p3,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/",grna,".pdf"),width = 10.5, height = 5)
}

rra_list <- c("Lgals1","Klf2","Gzmk","S100a10","Klrd1","Cd7","Cd160","Tcf7","Cd38","Lag3","Havcr2","Entpd1")
map(rra_list,rraplot)

```

#scmageck on pathway:GSEA
cp 23tpdeeper/path.txt to  23tpmerged/path.txt

#prepare signature files
#SIGNATURE = system.file("extdata", "test_symbols.gmt.txt", package = "scMAGeCK")
#sig <- read.delim("/data/msun/tools/R_lib/scMAGeCK/extdata/test_symbols.gmt.txt")
path1 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE9650_EFFECTOR_VS_MEMORY_CD8_TCELL_DN.v2023.1.Hs.gmt") 
path1$mg <- path1$gene %>% convert_human_to_mouse_symbols() 
path1 <- path1 %>% na.omit() %>% t()
path2 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE9650_NAIVE_VS_EXHAUSTED_CD8_TCELL_UP.v2023.1.Hs.gmt") 
path2$mg <- path2$gene %>% convert_human_to_mouse_symbols()
path2 <- path2 %>% na.omit() %>% t()
path3 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE9650_NAIVE_VS_EFF_CD8_TCELL_DN.v2023.1.Hs.gmt") 
path3$mg <- path3$gene %>% convert_human_to_mouse_symbols()
path3 <- path3 %>% na.omit() %>% t()
path4 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/KAECH_NAIVE_VS_DAY8_EFF_CD8_TCELL_DN.v2023.1.Hs.gmt") 
path4$mg <- path4$gene %>% convert_human_to_mouse_symbols()
path4 <- path4 %>% na.omit() %>% t()
path5 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE41978_KLRG1_HIGH_VS_LOW_EFFECTOR_CD8_TCELL_DN.v2023.1.Hs.gmt") 
path5$mg <- path5$gene %>% convert_human_to_mouse_symbols()
path5 <- path5 %>% na.omit() %>% t()
path6 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN.v2023.1.Hs.gmt") 
path6$mg <- path6$gene %>% convert_human_to_mouse_symbols()
path6 <- path6 %>% na.omit() %>% t()
path7 <- read.gmt("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/gmt/GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_UP.v2023.1.Hs.gmt") 
path7$mg <- path7$gene %>% convert_human_to_mouse_symbols()
path7 <- path7 %>% na.omit() %>% t()
path <- qpcR:::rbind.na(path1[c(1,3),],path2[c(1,3),],path3[c(1,3),],path4[c(1,3),],path5[c(1,3),],path6[c(1,3),],path7[c(1,3),])
write.csv(path,file = "/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/path.csv")#edit txt and upload
path <- read.delim("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/path.txt")#check its format

```{r}
library(R.utils)
library(nichenetr)
library(qpcR)
library(clusterProfiler)

pathway <- "/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/path.txt"
rra_gsea <- scmageck_rra(BARCODE=barcode, RDS=cl13_allgene, SIGNATURE = pathway,   
                            NEGCTRL="Scramble", KEEPTMP=FALSE) 

rragsea <- function(path){
  rra <- read.delim(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/GENE_SET/",path,"_RRA.txt")) %>%
              mutate(top10_low = rank(-log10(lo_value.low)) %in% 12:21) %>%
              mutate(top10_high = rank(-log10(lo_value.high)) %in% 12:21)
  p1 <- ggplot(data=rra,aes(x=reorder(Row.names,lo_value.low), y=-log10(lo_value.low),group=1)) +
    geom_line() +
    ggthemes::theme_few() +
    geom_point(data = rra[rra$FDR.low<0.05,],color="gray",size=2) +
    geom_point(data = rra[rra$top10_low,], color="#4197D9",size=2) +
    ggrepel::geom_text_repel(aes(label = ifelse(top10_low, as.character(Row.names), "")), size = 4,color="#4197D9") +
    xlab("cells with targeted gene") +
    ylab("-log10(RRA score)") +
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        plot.subtitle=element_text(size=10, face="italic")) +
    labs(title = "Gene KO enriched in effector score low",
              subtitle = str_c("gene set: ",path))
  p2 <- ggplot(data=rra,aes(x=reorder(Row.names,lo_value.high), y=-log10(lo_value.high),group=1)) +
    geom_line() +
    ggthemes::theme_few() +
    geom_point(data = rra[rra$top10_high,], color="orange",size=2) +
    ggrepel::geom_text_repel(aes(label = ifelse(top10_high, as.character(Row.names), "")), size = 4,color="orange") +
    xlab("cells with targeted gene") +
    ylab("-log10(RRA score)") +
    theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        plot.subtitle=element_text(size=10, face="italic")) +
    labs(title = "Gene KO enriched in effector score high",
              subtitle = str_c("gene set: ",path))
  p3 <- gridExtra::grid.arrange(p1, p2, ncol = 2)
  ggsave(p3,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/04_scmageck/",path,".pdf"),width = 10.5, height = 5)
}

rragsea("GSE9650_EFFECTOR_VS_MEMORY_CD8_TCELL_DN")
rragsea("GSE9650_NAIVE_VS_EXHAUSTED_CD8_TCELL_UP")
rragsea("GSE9650_NAIVE_VS_EFF_CD8_TCELL_DN")
rragsea("KAECH_NAIVE_VS_DAY8_EFF_CD8_TCELL_DN")
rragsea("GSE41978_KLRG1_HIGH_VS_LOW_EFFECTOR_CD8_TCELL_DN")
rragsea("GSE9650_EXHAUSTED_VS_MEMORY_CD8_TCELL_DN")
rragsea("GSE9650_EFFECTOR_VS_EXHAUSTED_CD8_TCELL_UP")
```

#deg of guide-ID
```{r}
genes <- c("Pdcd1","Cd160","Cd101","Cd244","Tigit","Cd38","Entpd1","Id2","Tox","Bhlhe41","Prdm1","Fil1","Ccr7","Il7r","Sell","Cd27","Cd28","Tcf7","Gzma","Gzmb","Prf1","Prdm1","Ikzf3","Jdp2","Hic1","Zfp324","stat3")

DEGvolcano <- function(sgRNA){
  deg <- FindMarkers(sub_cl13, ident.1 = sgRNA, ident.2 = "gScramble", test.use = "MAST",logfc.threshold = 0,only.pos = FALSE) %>% rownames_to_column(var = "names")
  write.csv(deg,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv"))

  DEGs <- deg %>% 
  mutate(group=case_when(
  p_val<0.05 & avg_log2FC>0 ~ "Upregulated",
  p_val<0.05 & avg_log2FC<0 ~ "Downregulated",
  TRUE ~ "No difference"
))
topgenes <- subset(DEGs, group %in% c("Upregulated","Downregulated")) %>%
  group_by(group) %>%
  top_n(n = 30, wt = abs(avg_log2FC)) %>%
  .$names

p <- ggplot(DEGs,aes(x=avg_log2FC,y=-log10(p_val))) +
  geom_point(size=1,aes(color=group)) +
  scale_color_manual(values = c("#4197D9", "grey", "orange")) +
  geom_vline(xintercept = 0.0,linetype=2) +
  geom_hline(yintercept = -log10(0.05),linetype=2) +
  theme_classic() +
  geom_text_repel(data = subset(DEGs, names %in% c(genes,topgenes)), aes(label = names), min.segment.length = 0, seed = 42, box.padding = 0.5,max.overlaps = 25) +
  ggtitle(str_c("DEG of ",sgRNA))

ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/", sgRNA,".pdf"),width = 9,height = 7)

}

safer_DEGvolcano <- possibly(DEGvolcano, otherwise = "Error in file") #skip the error files

Idents(sub_cl13) <- "perturbed_ID"
list <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
walk(list,safer_DEGvolcano)

Idents(sub_cl13) <- "perturbed_gene"
list <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
walk(list,safer_DEGvolcano)
```

#go analysis on immune pathways
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library(msigdbr)
m_t2g <- msigdbr(species = "Mus musculus", category = "C7") %>% 
  dplyr::select(gs_name, entrez_gene)

go <- function(sgRNA){
  DEG <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv")) %>% na.omit(.)
  upgenes <- DEG[DEG$p_val<0.05 & DEG$avg_log2FC>0,]$names %>%
    bitr(fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
  upgo <- enricher(gene=upgenes$ENTREZID, TERM2GENE=m_t2g)
  updf <- upgo@result %>% top_n(n = -30, wt = p.adjust)
  
  downgenes <- DEG[DEG$p_val<0.05 & DEG$avg_log2FC<0,]$names %>%
    bitr(fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
  downgo <- enricher(gene=downgenes$ENTREZID, TERM2GENE=m_t2g)
  downdf <- downgo@result %>% top_n(n = -30, wt = p.adjust)
  
  gotable <- downdf %>% mutate(Count=-Count) %>% rbind(updf) %>% subset(.,p.adjust<0.05)
  write.csv(gotable,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/go_immune/",sgRNA,".csv"))
  
p <- ggplot(gotable,aes(x = Description,y = Count))+
    geom_bar(stat="identity", aes(fill = p.adjust)) + 
    coord_flip() + 
    scale_fill_continuous(limits = c(0,0.05), low="red", high="blue",guide=guide_colorbar(reverse=TRUE)) +
    scale_x_discrete(limits=gotable$Description) +  
    theme_minimal() + 
    ggtitle(str_c("GO analysis of ",sgRNA))

ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/go_immune/", sgRNA,".pdf"),width = 15,height = 10)
}

safer_go <- possibly(go, otherwise = "Error in file")
list <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
walk(list,safer_go)

list <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
walk(list,safer_go)
```

#GSEA on immune pathways
```{r}
gseaplot <- function(sgRNA){
  DEG <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv")) %>% na.omit(.)
  id <- bitr(DEG$names,fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db") 
  DEG <- merge(id,DEG,by.x = "SYMBOL", by.y = "names",all.x=TRUE)
  genelist <- as.numeric(DEG$avg_log2FC)
  names(genelist) = as.character(DEG$ENTREZID)
  genelist = sort(genelist, decreasing = TRUE)
  gse <- GSEA(genelist, TERM2GENE = m_t2g,pvalueCutoff = 1)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_immune/",sgRNA,".csv"))
  
  p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
    facet_grid(.~.sign) +
    ggtitle(str_c("GSEA analysis of ",sgRNA))
  ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_immune/", sgRNA,".pdf"),width = 18,height = 14) 
}

safer_gseaplot <- possibly(gseaplot, otherwise = "Error in file")

list <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
walk(list,safer_gseaplot)
#gZscan20-4 has no term enriched under specific pvalueCutoff

list <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
walk(list,safer_gseaplot)
#gZscan20 has no term enriched under specific pvalueCutoff
```

#go/gsea on MF pathways (molecular function)
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library(msigdbr)
mf <- msigdbr(species = "Mus musculus", category = "C5",subcategory = "MF") %>% 
  dplyr::select(gs_name, entrez_gene)

go <- function(sgRNA){
  DEG <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv")) %>% na.omit(.)
  upgenes <- DEG[DEG$p_val<0.05 & DEG$avg_log2FC>0,]$names %>%
    bitr(fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
  upgo <- enricher(gene=upgenes$ENTREZID, TERM2GENE=mf)
  updf <- upgo@result %>% top_n(n = -30, wt = p.adjust)
  
  downgenes <- DEG[DEG$p_val<0.05 & DEG$avg_log2FC<0,]$names %>%
    bitr(fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
  downgo <- enricher(gene=downgenes$ENTREZID, TERM2GENE=mf)
  downdf <- downgo@result %>% top_n(n = -30, wt = p.adjust)
  
  gotable <- downdf %>% mutate(Count=-Count) %>% rbind(updf) %>% subset(.,p.adjust<0.05)
  write.csv(gotable,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/go_mf/",sgRNA,".csv"))
  
p <- ggplot(gotable,aes(x = Description,y = Count))+
    geom_bar(stat="identity", aes(fill = p.adjust)) + 
    coord_flip() + 
    scale_fill_continuous(limits = c(0,0.05), low="red", high="blue",guide=guide_colorbar(reverse=TRUE)) +
    scale_x_discrete(limits=gotable$Description) +  
    theme_minimal() + 
    ggtitle(str_c("GO analysis of ",sgRNA))

ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/go_mf/", sgRNA,".pdf"),width = 15,height = 10)
}

safer_go <- possibly(go, otherwise = "Error in file")
list <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
walk(list,safer_go)
# gNfatc1-4, gZscan20-4, gIrf8-4, gStat3-4,gTfdp1-2 have no term enriched

list <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
walk(list,safer_go)
#gZfp410 Nfatc1 Irf8 have no terms enriched

gseaplot <- function(sgRNA){
  DEG <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv")) %>% na.omit(.)
  id <- bitr(DEG$names,fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db") 
  DEG <- merge(id,DEG,by.x = "SYMBOL", by.y = "names",all.x=TRUE)
  genelist <- as.numeric(DEG$avg_log2FC)
  names(genelist) = as.character(DEG$ENTREZID)
  genelist = sort(genelist, decreasing = TRUE)
  gse <- GSEA(genelist, TERM2GENE = mf)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_mf/",sgRNA,".csv"))
  
  p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
    facet_grid(.~.sign) +
    ggtitle(str_c("GSEA analysis of ",sgRNA))
  ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_mf/", sgRNA,".pdf"),width = 18,height = 14) 
}

safer_gseaplot <- possibly(gseaplot, otherwise = "Error in file")
list <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
walk(list,safer_gseaplot)

list <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
walk(list,safer_gseaplot)
#gZscan20 has no term enriched under specific pvalueCutoff
```


# intersection of up/down DEGs
```{r}
sub_cl13 <- readRDS("/data/msun/projects/kay_ECCITE/23TP_merged/sub_cl13.rds")
#intersection of up DEG tables
library(openxlsx)
updeg <- function(sgrna){
  df=read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgrna,".csv"))
  deg <- df %>% mutate(group=case_when(p_val<0.05 & avg_log2FC>0 ~ "Upregulated",
                         p_val<0.05 & avg_log2FC<0 ~ "Downregulated",
                         TRUE ~ "No difference"))
  topgenes <- subset(deg, group %in% c("Upregulated")) %>%
    top_n(n = 50, wt = avg_log2FC) %>%
    .$names
  topgenes
}
list1 <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
list2 <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble'] 
list <- c(list1,list2)
safer_updeg <- possibly(updeg, otherwise = "Error in file")
upDEG <- map(list,safer_updeg) 
names(upDEG) <- list

df2 <- data.frame(gene=unique(unlist(upDEG)))
df1 <- lapply(upDEG,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "guide_ID")

df_int <- lapply(df2$gene,function(x){
  # pull the name of the intersections
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(guide_ID) %>% 
    pull("guide_ID") %>% 
    paste0(collapse = "|")
  # build the dataframe
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()
df_int$count <- str_count(df_int$int,"-")
write.csv(df_int,file="/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/upDEG_intersection_cl13.csv")

#intersection of Down DEG tables-------
downdeg <- function(sgrna){
  df=read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgrna,".csv"))
  deg <- df %>% mutate(group=case_when(p_val<0.05 & avg_log2FC>0 ~ "Upregulated",
                         p_val<0.05 & avg_log2FC<0 ~ "Downregulated",
                         TRUE ~ "No difference"))
  topgenes <- subset(deg, group %in% c("Downregulated")) %>%
    top_n(n = 50, wt = abs(avg_log2FC)) %>%
    .$names
  topgenes
}

safer_downdeg <- possibly(downdeg, otherwise = "Error in file")
downDEG <- map(list,safer_downdeg) 
names(downDEG) <- list

df2 <- data.frame(gene=unique(unlist(downDEG)))
df1 <- lapply(downDEG,function(x){
  data.frame(gene = x)
}) %>% 
  bind_rows(.id = "guide_ID")

df_int <- lapply(df2$gene,function(x){
  # pull the name of the intersections
  intersection <- df1 %>% 
    dplyr::filter(gene==x) %>% 
    arrange(guide_ID) %>% 
    pull("guide_ID") %>% 
    paste0(collapse = "|")
  # build the dataframe
  data.frame(gene = x,int = intersection)
}) %>% 
  bind_rows()
df_int$count <- str_count(df_int$int,"-")
write.csv(df_int,file="/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/downDEG_intersection_cl13.csv")

#upset plots 
#library(UpSetR)
#upset(fromList(upDEG),nsets=15) 
```


# intersection of pathways : merge all the results 
```{r}
#merge go results ------------------
intersection_go <- function(sgrna,pathway){
  alldf=read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/go_",pathway,"/",sgrna,".csv"))
  alldf=as.data.frame(alldf[,colnames(alldf) %in% c("ID","GeneRatio","p.adjust","Count")]) #count indicate up or down regulated
  colnames(alldf)=c("path",str_c("GeneRatio_",sgrna),str_c("adjP_",sgrna),str_c("regulated_",sgrna))
  alldf[which(alldf[,3] < 0.05 & alldf[,4] > 0),4] <- 'Up'
  alldf[which(alldf[,3] < 0.05 & alldf[,4] < 0),4] <- 'Down'
  alldf[which(alldf[,3] >= 0.05),4] <- 'None'
  alldf=na.omit(alldf) %>% 
    mutate_if(is.logical, as.character)
  alldf
}
list1 <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
list2 <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
list <- c(list1,list2)
safer_intersection_go <- possibly(intersection_go, otherwise = "NA")

halflist<- map2(list,"mf",safer_intersection_go) %>%
  .[. != "NA"] 
#append dataframe of pathname to the list 
pathdf <- halflist %>% 
  purrr::reduce(.,full_join,by="path") 
pathdf <- as.data.frame(unique(pathdf$path))
colnames(pathdf) <- "path"
alllist <- append(list(pathdf),halflist)
all_go_mf <- purrr::reduce(alllist,left_join,by="path")
write.csv(all_go_mf,file = "/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gomf_intersection_cl13.csv")

halflist<- map2(list,"immune",safer_intersection_go) %>%
  .[. != "NA"] 
#append dataframe of pathname to the list 
pathdf <- halflist %>% 
  purrr::reduce(.,full_join,by="path") 
pathdf <- as.data.frame(unique(pathdf$path))
colnames(pathdf) <- "path"
alllist <- append(list(pathdf),halflist)
all_go_immune <- purrr::reduce(alllist,left_join,by="path")
write.csv(all_go_immune,file = "/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/goimmune_intersection_cl13.csv")


#merge gsea results-------------------------------
intersection_gsea <- function(sgrna,pathway){
  alldf=read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_",pathway,"/",sgrna,".csv"))
  alldf=as.data.frame(alldf[,colnames(alldf) %in% c("ID","NES","p.adjust")]) #count indicate up or down regulated
  colnames(alldf)=c("path",str_c("NES_",sgrna),str_c("adjP_",sgrna))
  alldf[which(alldf[,3] < 0.05 & alldf[,2] > 0),str_c("regulated_",sgrna)] <- 'Up'
  alldf[which(alldf[,3] < 0.05 & alldf[,2] < 0),str_c("regulated_",sgrna)] <- 'Down'
  alldf[which(alldf[,3] >= 0.05),str_c("regulated_",sgrna)] <- 'None'
  alldf=na.omit(alldf) %>% 
    mutate_if(is.logical, as.character)
  alldf
}
list1 <- unique(sub_cl13$perturbed_ID) %>% .[ !. == 'gScramble'] 
list2 <- unique(sub_cl13$perturbed_gene) %>% .[ !. == 'gScramble']
list <- c(list1,list2)
safer_intersection_gsea <- possibly(intersection_gsea, otherwise = "NA")

halflist<- map2(list,"mf",safer_intersection_gsea) %>%
  .[. != "NA"] 
#append dataframe of pathname to the list 
pathdf <- halflist %>% 
  purrr::reduce(.,full_join,by="path") 
pathdf <- as.data.frame(unique(pathdf$path))
colnames(pathdf) <- "path"
alllist <- append(list(pathdf),halflist)
all_gsea_mf <- purrr::reduce(alllist,left_join,by="path")
write.csv(all_gsea_mf,file = "/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gseamf_intersection_cl13.csv")

halflist<- map2(list,"immune",safer_intersection_gsea) %>%
  .[. != "NA"] 
#append dataframe of pathname to the list 
pathdf <- halflist %>% 
  purrr::reduce(.,full_join,by="path") 
pathdf <- as.data.frame(unique(pathdf$path))
colnames(pathdf) <- "path"
alllist <- append(list(pathdf),halflist)
all_gsea_immune <- purrr::reduce(alllist,left_join,by="path")
write.csv(all_gsea_immune,file = "/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gseaimmune_intersection_cl13.csv")

```

#run gsea on c5-c7 and hallmark
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library(msigdbr)
c7 <- msigdbr(species = "Mus musculus", category = "C7") %>% 
  dplyr::select(gs_name, entrez_gene)
mf <- msigdbr(species = "Mus musculus", category = "C5",subcategory = "MF") %>% 
  dplyr::select(gs_name, entrez_gene)
c6 <- msigdbr(species = "Mus musculus", category = "C6") %>% 
  dplyr::select(gs_name, entrez_gene)
hallmark <- msigdbr(species = "Mus musculus", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

gogsea <- function(sgRNA){
  
  DEG <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/DEG_volcano/",sgRNA,".csv")) %>% na.omit(.)
  id <- bitr(DEG$names,fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db") 
  DEG <- merge(id,DEG,by.x = "SYMBOL", by.y = "names",all.x=TRUE)
  genelist <- as.numeric(DEG$avg_log2FC)
  names(genelist) = as.character(DEG$ENTREZID)
  genelist = sort(genelist, decreasing = TRUE)
  
#C7-------
  gse <- GSEA(genelist, TERM2GENE = c7,pvalueCutoff = 1)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_immune/",sgRNA,".csv"))
  
 p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
   facet_grid(.~.sign) +
   ggtitle(str_c("GSEA analysis of ",sgRNA))
   ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_immune/",sgRNA,".pdf"),width = 18,height = 14) 
  
#mf----------
  gse <- GSEA(genelist, TERM2GENE = mf,pvalueCutoff = 1)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_mf/",sgRNA,".csv"))
  
 p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
   facet_grid(.~.sign) +
   ggtitle(str_c("GSEA analysis of ",sgRNA))
   ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_mf/", sgRNA,".pdf"),width = 18,height = 14) 
   
#c6----------
  gse <- GSEA(genelist, TERM2GENE = c6,pvalueCutoff = 1)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_c6/",sgRNA,".csv"))
  
 p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
   facet_grid(.~.sign) +
   ggtitle(str_c("GSEA analysis of ",sgRNA))
   ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_c6/", sgRNA,".pdf"),width = 18,height = 14) 
   
#hallmark----------
  gse <- GSEA(genelist, TERM2GENE = hallmark,pvalueCutoff = 1)
  write.csv(gse,file = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_hallmark/",sgRNA,".csv"))
  
 p <- dotplot(gse, showCategory=30, split=".sign", x="NES",label_format = 100) + 
   facet_grid(.~.sign) +
   ggtitle(str_c("GSEA analysis of ",sgRNA))
   ggsave(p,filename = str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_hallmark/", sgRNA,".pdf"),width = 18,height = 14) 
}

safer_gogsea <- possibly(gogsea, otherwise = "Error in file")

list1 <- unique(obj$perturbed_ID) %>% .[ !. == 'gScramble'] 
list2 <- unique(obj$perturbed_gene) %>% .[ !. == 'gScramble']
list <- c(list1,list2)
walk(list,safer_gogsea)

intersection_gsea <- function(sgrna,obj){
  mf_table <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_mf/",sgrna,".csv"))
  c6_table <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_c6/",sgrna,".csv")) 
  immune_table <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_immune/",sgrna,".csv"))
  hallmark_table <- read.csv(str_c("/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_hallmark/",sgrna,".csv"))
  alldf <- do.call("rbind", list(mf_table, c6_table, immune_table,hallmark_table))
  alldf=as.data.frame(alldf[,colnames(alldf) %in% c("ID","NES","p.adjust")]) #count indicate up or down regulated
  colnames(alldf)=c("path",str_c("NES_",sgrna),str_c("adjP_",sgrna))
  alldf[which(alldf[,3] < 0.05 & alldf[,2] > 0),str_c("regulated_",sgrna)] <- 'Up'
  alldf[which(alldf[,3] < 0.05 & alldf[,2] < 0),str_c("regulated_",sgrna)] <- 'Down'
  alldf[which(alldf[,3] >= 0.05),str_c("regulated_",sgrna)] <- 'None'
  alldf=na.omit(alldf) %>% 
    mutate_if(is.logical, as.character)
  alldf
}
safer_intersection_gsea <- possibly(intersection_gsea, otherwise = "NA")

datalist<- map2(list,"Arm_SPL",safer_intersection_gsea) %>%
  .[. != "NA"] %>% 
  purrr::reduce(.,merge,by="path") 
write.csv(datalist,file = "/data/msun/projects/kay_ECCITE/23TP_merged/05_DEG_mast/gsea_intersection_cl13from23TP.csv")

```
